tools/cli/adminDBCommands.go

```diff
@@ -0,0 +1,631 @@
+	executionsPageSize := c.Int(FlagPageSize)
+	scanReportRate := c.Int(FlagScanReportRate)
+
+	payloadSerializer := persistence.NewPayloadSerializer()
+	rateLimiter := getRateLimiter(startingRPS, targetRPS, scaleUpSeconds)
+	session := connectToCassandra(c)
+	historyStore := cassp.NewHistoryV2PersistenceFromSession(session, loggerimpl.NewNopLogger())
+	branchDecoder := codec.NewThriftRWEncoder()
+	scanOutputDirectories := createScanOutputDirectories()
+	defer session.Close()
+
+	reports := make(chan *ShardScanReport)
+	for i := 0; i < scanWorkerCount; i++ {
+		go func(workerIdx int) {
+			for shardID := 0; shardID < numShards; shardID++ {
+				if shardID%scanWorkerCount == workerIdx {
+					reports <- scanShard(
+						session,
+						shardID,
+						scanOutputDirectories,
+						rateLimiter,
+						executionsPageSize,
+						payloadSerializer,
+						historyStore,
+						branchDecoder)
+				}
+			}
+		}(i)
+	}
+
+	progressReport := &ProgressReport{}
+	for i := 0; i < numShards; i++ {
+		report := <-reports
+		includeShardInProgressReport(report, progressReport)
+		if i%scanReportRate == 0 {
+			reportBytes, err := json.MarshalIndent(*progressReport, "", "\t")
+			if err != nil {
+				ErrorAndExit("failed to print progress", err)
+			}
+			fmt.Println(string(reportBytes))
+		}
+	}
+}
+
+func scanShard(
+	session *gocql.Session,
+	shardID int,
+	scanOutputDirectories *ScanOutputDirectories,
+	limiter *quotas.DynamicRateLimiter,
+	executionsPageSize int,
+	payloadSerializer persistence.PayloadSerializer,
+	historyStore persistence.HistoryStore,
+	branchDecoder *codec.ThriftRWEncoder,
+) *ShardScanReport {
+	outputFiles, closeFn := createShardScanOutputFiles(shardID, scanOutputDirectories)
+	report := &ShardScanReport{
+		ShardID: shardID,
+	}
+	defer func() {
+		recordShardScanReport(outputFiles.ShardScanReportFile, report)
+		deleteEmptyShardScanOutputFiles(outputFiles)
+		closeFn()
+	}()
+	execStore, err := cassp.NewWorkflowExecutionPersistence(shardID, session, loggerimpl.NewNopLogger())
+	if err != nil {
+		report.Failure = &ShardScanReportFailure{
+			Note:    "failed to create execution store",
+			Details: err.Error(),
+		}
+		return report
+	}
+
+	var token []byte
+	isFirstIteration := true
+	for isFirstIteration || len(token) != 0 {
+		isFirstIteration = false
+		req := &persistence.ListConcreteExecutionsRequest{
+			PageSize:  executionsPageSize,
+			PageToken: token,
+		}
+		limiter.Wait(context.Background())
+		resp, err := execStore.ListConcreteExecutions(req)
+		if err != nil {
+			report.Failure = &ShardScanReportFailure{
+				Note:    "failed to call ListConcreteExecutions",
+				Details: err.Error(),
+			}
+			return report
+		}
+		token = resp.NextPageToken
+		for _, e := range resp.ExecutionInfos {
+			if report.Scanned == nil {
+				report.Scanned = &ShardScanReportExecutionsScanned{}
+			}
+			report.Scanned.TotalExecutionsCount++
+			historyVerificationResult, history, historyBranch := verifyHistoryExists(
```

Comment:
Doesn't have to be in this PR. But we can have something like plugin mode. Easy to add new checks and from the CLI, we can specify which check to execute. What do you think?

Reply:
yes totally agree I will do this as part of scanner though. CLI can consume this once we have it in workflow.